File: go/gateway.go
package owlbot

import (
	"encoding/json"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// Gateway 实时事件连接：HELLO/IDENTIFY/HEARTBEAT 自动处理，断线指数退避重连。
type Gateway struct {
	client *Client

	mu       sync.Mutex
	handlers map[string][]func(json.RawMessage)
	closed   bool
	conn     *websocket.Conn
}

type gatewayFrame struct {
	Op string          `json:"op"`
	T  string          `json:"t,omitempty"`
	D  json.RawMessage `json:"d,omitempty"`
}

// ConnectGateway 连接 Gateway；用 On 注册事件回调后事件即开始分发。
func (c *Client) ConnectGateway() *Gateway {
	g := &Gateway{client: c, handlers: make(map[string][]func(json.RawMessage))}
	go g.run()
	return g
}

// On 注册事件回调：事件名如 MESSAGE_CREATE / MESSAGE_STREAM_DELTA，
// 内置事件 "READY"（连接就绪）。回调收到原始 JSON 载荷。
func (g *Gateway) On(event string, handler func(payload json.RawMessage)) *Gateway {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.handlers[event] = append(g.handlers[event], handler)
	return g
}

// Close 关闭连接并停止重连。
func (g *Gateway) Close() {
	g.mu.Lock()
	g.closed = true
	conn := g.conn
	g.mu.Unlock()
	if conn != nil {
		_ = conn.Close()
	}
}

func (g *Gateway) emit(event string, payload json.RawMessage) {
	g.mu.Lock()
	handlers := append([]func(json.RawMessage){}, g.handlers[event]...)
	g.mu.Unlock()
	for _, handler := range handlers {
		handler(payload)
	}
}

func (g *Gateway) wsURL() string {
	base := g.client.apiBase
	if strings.HasPrefix(base, "https://") {
		base = "wss://" + strings.TrimPrefix(base, "https://")
	} else {
		base = "ws://" + strings.TrimPrefix(base, "http://")
	}
	return base + "/gateway"
}

func (g *Gateway) run() {
	backoff := time.Second
	for {
		g.mu.Lock()
		if g.closed {
			g.mu.Unlock()
			return
		}
		g.mu.Unlock()

		if g.session() {
			backoff = time.Second // 会话曾就绪：重置退避
		}

		g.mu.Lock()
		closed := g.closed
		g.mu.Unlock()
		if closed {
			return
		}
		time.Sleep(backoff)
		if backoff < 30*time.Second {
			backoff *= 2
		}
	}
}

// session 执行一次完整连接生命周期；返回是否曾进入 READY。
func (g *Gateway) session() bool {
	conn, _, err := websocket.DefaultDialer.Dial(g.wsURL(), nil)
	if err != nil {
		return false
	}
	g.mu.Lock()
	g.conn = conn
	g.mu.Unlock()
	defer conn.Close()

	ready := false
	stopBeat := make(chan struct{})
	defer close(stopBeat)

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return ready
		}
		var frame gatewayFrame
		if err := json.Unmarshal(raw, &frame); err != nil {
			continue
		}
		switch frame.Op {
		case "HELLO":
			identify, _ := json.Marshal(map[string]any{
				"op": "IDENTIFY", "d": map[string]string{"token": g.client.token},
			})
			if err := conn.WriteMessage(websocket.TextMessage, identify); err != nil {
				return ready
			}
			var hello struct {
				HeartbeatIntervalMS int64 `json:"heartbeat_interval_ms"`
			}
			_ = json.Unmarshal(frame.D, &hello)
			interval := time.Duration(hello.HeartbeatIntervalMS) * time.Millisecond
			if interval <= 0 {
				interval = 30 * time.Second
			}
			go g.heartbeatLoop(conn, interval, stopBeat)
		case "READY":
			ready = true
			g.emit("READY", frame.D)
		case "DISPATCH":
			g.emit(frame.T, frame.D)
		}
	}
}

func (g *Gateway) heartbeatLoop(conn *websocket.Conn, interval time.Duration, stop <-chan struct{}) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	beat, _ := json.Marshal(map[string]string{"op": "HEARTBEAT"})
	for {
		select {
		case <-stop:
			return
		case <-ticker.C:
			if err := conn.WriteMessage(websocket.TextMessage, beat); err != nil {
				return
			}
		}
	}
}

File: rust/src/gateway.rs
//! Gateway 实时事件：HELLO / IDENTIFY / HEARTBEAT，断线指数退避重连。

use crate::client::Client;
use crate::error::{Error, Result};
use crate::interaction::Interaction;
use futures_util::{SinkExt, StreamExt};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{mpsc, Mutex};
use tokio_tungstenite::{connect_async, tungstenite::Message};

/// 事件回调：`(事件名, 载荷 JSON)`。内置事件 `"READY"` 表示连接就绪。
pub type EventHandler = Arc<dyn Fn(String, Value) + Send + Sync + 'static>;

/// Gateway 连接句柄。
pub struct Gateway {
    stop_tx: mpsc::Sender<()>,
    join: tokio::task::JoinHandle<()>,
}

impl Gateway {
    /// 关闭连接并停止重连。
    pub async fn close(self) {
        let _ = self.stop_tx.send(()).await;
        let _ = self.join.await;
    }
}

/// 线程安全的事件处理表。
#[derive(Default, Clone)]
pub struct HandlerMap {
    inner: Arc<Mutex<HashMap<String, Vec<EventHandler>>>>,
}

impl HandlerMap {
    pub fn new() -> Self {
        Self::default()
    }

    /// 注册事件回调（如 `MESSAGE_CREATE` / `MESSAGE_REACTION_ADD` / `READY`）。
    pub async fn on<F>(&self, event: impl Into<String>, handler: F)
    where
        F: Fn(String, Value) + Send + Sync + 'static,
    {
        let mut map = self.inner.lock().await;
        map.entry(event.into())
            .or_default()
            .push(Arc::new(handler));
    }

    /// 注册按钮点击回调：`INTERACTION_CREATE` 载荷自动包装成 [`Interaction`]。
    ///
    /// 需要传入 `Client`（用于回应 callback 端点）；解析失败的载荷会被忽略。
    pub async fn on_interaction<F>(&self, client: Client, handler: F)
    where
        F: Fn(Interaction) + Send + Sync + 'static,
    {
        self.on("INTERACTION_CREATE", move |_, payload| {
            if let Ok(interaction) = Interaction::from_value(client.clone(), payload) {
                handler(interaction);
            }
        })
        .await;
    }

    async fn emit(&self, event: &str, payload: Value) {
        let handlers = {
            let map = self.inner.lock().await;
            map.get(event).cloned().unwrap_or_default()
        };
        for h in handlers {
            h(event.to_string(), payload.clone());
        }
    }
}

/// 连接 Gateway 并开始分发事件。
pub async fn connect(client: Client, handlers: HandlerMap) -> Result<Gateway> {
    let (stop_tx, mut stop_rx) = mpsc::channel::<()>(1);
    let handlers = Arc::new(handlers);
    let client = Arc::new(client);

    let join = tokio::spawn(async move {
        let mut backoff_ms: u64 = 1000;
        loop {
            tokio::select! {
                _ = stop_rx.recv() => break,
                result = run_session(client.clone(), handlers.clone()) => {
                    match result {
                        Ok(true) => backoff_ms = 1000,
                        Ok(false) | Err(_) => {}
                    }
                }
            }
            tokio::select! {
                _ = stop_rx.recv() => break,
                _ = tokio::time::sleep(std::time::Duration::from_millis(backoff_ms)) => {
                    backoff_ms = (backoff_ms * 2).min(30_000);
                }
            }
        }
    });

    Ok(Gateway { stop_tx, join })
}

async fn run_session(client: Arc<Client>, handlers: Arc<HandlerMap>) -> Result<bool> {
    let ws_url = gateway_url(client.api_base());
    let (ws, _) = connect_async(&ws_url)
        .await
        .map_err(|e| Error::WebSocket(e.to_string()))?;
    let (write, mut read) = ws.split();
    let write = Arc::new(Mutex::new(write));

    let mut ready = false;
    let mut heartbeat_task: Option<tokio::task::JoinHandle<()>> = None;

    while let Some(msg) = read.next().await {
        let msg = msg.map_err(|e| Error::WebSocket(e.to_string()))?;
        let text = match msg {
            Message::Text(t) => t.to_string(),
            Message::Binary(b) => String::from_utf8_lossy(&b).into_owned(),
            Message::Close(_) => break,
            Message::Ping(p) => {
                let mut w = write.lock().await;
                let _ = w.send(Message::Pong(p)).await;
                continue;
            }
            _ => continue,
        };
        let frame: Value = match serde_json::from_str(&text) {
            Ok(v) => v,
            Err(_) => continue,
        };
        let op = frame.get("op").and_then(|v| v.as_str()).unwrap_or("");
        match op {
            "HELLO" => {
                let identify = serde_json::json!({
                    "op": "IDENTIFY",
                    "d": { "token": client.token() }
                });
                {
                    let mut w = write.lock().await;
                    w.send(Message::Text(identify.to_string().into()))
                        .await
                        .map_err(|e| Error::WebSocket(e.to_string()))?;
                }
                let interval_ms = frame
                    .get("d")
                    .and_then(|d| d.get("heartbeat_interval_ms"))
                    .and_then(|v| v.as_u64())
                    .unwrap_or(30_000);
                if let Some(task) = heartbeat_task.take() {
                    task.abort();
                }
                let write_hb = write.clone();
                heartbeat_task = Some(tokio::spawn(async move {
                    let mut ticker =
                        tokio::time::interval(std::time::Duration::from_millis(interval_ms));
                    // 跳过第一次立即触发
                    ticker.tick().await;
                    loop {
                        ticker.tick().await;
                        let beat = Message::Text(
                            serde_json::json!({ "op": "HEARTBEAT" }).to_string().into(),
                        );
                        let mut w = write_hb.lock().await;
                        if w.send(beat).await.is_err() {
                            break;
                        }
                    }
                }));
            }
            "READY" => {
                ready = true;
                let d = frame.get("d").cloned().unwrap_or(Value::Null);
                handlers.emit("READY", d).await;
            }
            "DISPATCH" => {
                let t = frame
                    .get("t")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string();
                let d = frame.get("d").cloned().unwrap_or(Value::Null);
                if !t.is_empty() {
                    handlers.emit(&t, d).await;
                }
            }
            "HEARTBEAT_ACK" => {}
            _ => {}
        }
    }

    if let Some(task) = heartbeat_task {
        task.abort();
    }
    Ok(ready)
}

fn gateway_url(api_base: &str) -> String {
    let base = if let Some(rest) = api_base.strip_prefix("https://") {
        format!("wss://{rest}")
    } else if let Some(rest) = api_base.strip_prefix("http://") {
        format!("ws://{rest}")
    } else {
        api_base.to_string()
    };
    format!("{base}/gateway")
}
